uwblueprint/community-fridge-kw
# fix: custom email handler for firebase user verification

## Notion ticket link

[Fix logout](https://www.notion.so/uwblueprintexecs/Fix-logout-92188135087f43388e3df7e0a7035d65)



## Implementation description
The reason why the "logout" capability was not working was because the middleware check for the service, `isAuthorizedByUserId` was failing (returning false). Why? 

```typescript
      const firebaseUser = await firebaseAdmin
        .auth()
        .getUser(decodedIdToken.uid);
``` 
the firebase user object always had its `emailVerified` property as false. This means that for some reason, the email verification link sent on sign up was not working.
It works if we change it to the default:

![Screen Shot 2021-12-20 at 10 18 58 PM](https://user-images.githubusercontent.com/19617248/146865310-aabb26be-b0de-43ab-bf0e-c2e9dd1d72b0.png)
Somewhere along the way it got changed to `localhost:3000/login` which was why the link didn't work. This gives us: 
![Screen Shot 2021-12-20 at 10 20 32 PM](https://user-images.githubusercontent.com/19617248/146865460-98230e6e-9c48-4600-8e1d-4d142615955b.png)

However, we want to have a customized page and email handler for our use case for community fridge. A verification link consists of the form below
`http://localhost:3000/action?mode=verifyEmail&oobCode=5p567nctXDlRny89aSLNGIsN_H8NHlfGUJQCQnutGCQAAAF92uh2IQ&apiKey=AIzaSyCUNHAOONb1RFL4Do8EXc8Ox3Q3VxcTvGY&lang=en`

Most importantly, we want the `oobCode` which is a one-time unique code to verify our request. `Firebase-admin` does not expose a `applyActionCode` that `firebase` does which we could have applied easily to verify the email like [this](https://github.com/griffinsockwell/react-firebase-custom-email-handlers/blob/00928d88baff5efcb6d6084357571d0a828c0a15/src/components/VerifyEmail.js#L15)

Instead, we need to use the firebase Auth Rest API to verify the email code, ([See instructions here](https://firebase.google.com/docs/reference/rest/auth/#section-confirm-email-verification)) So now, when the user clicks on the custom action url`localhost:3000/action` from the email, we take the `oobCode` from the param and send a request to firebase to verify.

This sets the user record to have `emailVerified` as true and signout to work as expected!

**Example walkthrough:** 

https://user-images.githubusercontent.com/19617248/146864691-5999d5b0-bdcb-454d-8048-daed59b2e7b2.mov


## Steps to test
1. Pull branch, signup with a new user email that you haven't used before and have access to 
2. Click on the "confirm email verification link" sent to your email
3. Check that you can log in with this email and **logout** from the platform successfully

**Todo**
- [X] Add in appropriate error UI for if the code is invalid, been used, etc.
![Screen Shot 2021-12-20 at 11 25 36 PM](https://user-images.githubusercontent.com/19617248/146870890-53c00a93-e36f-48be-af2c-4bf9abd6539c.png)
